### PR TITLE
Integration cb httpserver

### PR DESCRIPTION
**Pull request by jicarretero**
* Integrated with CB and set up working Dockerfile for the server - Docker generated has been tested
* Changed the way jsonld are generated according to the way they are newly generated in transform.save()

- Tech. debt: -- Generation code for jsonld is coded 3 times. We should stop and refactor that.

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -1,63 +1,44 @@
...
+ARG INSTALLATION_PATH=/opt/$COMPONENT
 
-FROM python:3.9-alpine@sha256:e80214a705236091ee3821a7e512e80bd3337b50a95392a36b9a40b8fc0ea183
+RUN mkdir -p $INSTALLATION_PATH
+COPY --from=git-clone $INSTALLATION_PATH/requirements.txt /requirements.txt
+RUN pip install --root-user-action=ignore --prefix=$INSTALLATION_PATH -r /requirements.txt
 
-ENV PORT=${IOTA_PORT:-5000}
-
-LABEL "maintainer"="FIWARE Foundation e.V."
```
Comment (flopezag) [inline docker/Dockerfile:33]: why did you delete this information, at least in the last layer we should provide this information but by checking the content we should modify some of them:

LABEL "description"="An Internet of Things Agent for the RDF Turtle of SDMX DataQube format. This IoT Agent is designed to be a bridge between SDMX statistical metadata representation in RDF Turtle and  ETSI NGSI-LD in JSON-LD format representation to be integrated with FIWARE components."

LABEL "summary"="IoT Agent for the DataQube (SDMX) representation in RDF Turtle format used in statistical environments."

LABEL "org.opencontainers.image.documentation"="https://github.com/flopezag/IoTAgent-Turtle/tree/master/doc"

LABEL "org.opencontainers.image.description"="An Internet of Things Agent for the RDF Turtle of SDMX DataQube format. This IoT Agent is designed to be a bridge between SDMX statistical metadata representation in RDF Turtle and  ETSI NGSI-LD in JSON-LD format representation to be integrated with FIWARE components."

LABEL "org.python.version"="python:3.11"

File: docker/Readme.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
+# Docker for server
+
+## build
+The docker is easyly build using the following command - By default it will build the docker for the **develop** branch:
```
Comment (flopezag) [inline docker/Readme.md:4]: The docker is easily built using the following command - By default, it builds the docker for the **develop** branch:

File: docker/Dockerfile [hunk 3]
```diff
@@ -1,63 +1,44 @@
-FROM  python:3.9-alpine@sha256:e80214a705236091ee3821a7e512e80bd3337b50a95392a36b9a40b8fc0ea183  as  build
+FROM alpine/git as git-clone
 
-LABEL "maintainer"="FIWARE Foundation e.V."
-LABEL "description"="An Internet of Things Agent for the RDF DataQube Turtle format. This IoT Agent is designed to be a bridge between ISOXML/ADAPT and the ETSI NGSI-LD interface of a FIWARE Context Broker."
-LABEL "name"="iotagent-turtle"
-LABEL "summary"="IoT Agent for the DataQube representation in RDF Turtle format used in statistical environments"
+ARG PROJECT=flopezag
+ARG COMPONENT=IoTAgent-Turtle
+ARG VERSION=develop
```
Comment (flopezag) [inline docker/Dockerfile:5]: This should be BRANCH not VERSION

File: docker/Readme.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+```
+ docker build . -t iotagent-turtle 
+```
+
+If we want to build other branch, we could define the branch we want to use this way:
+```
+ docker build . -t iotagent-turtle --build-arg VERSION=integration-cb-httpserver
```
Comment (flopezag) [inline docker/Readme.md:12]: Following the previous comment, I suggest changing VERSION by BRANCH

File: docker/Readme.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    "level": "debug",
+    "rotation": "20 days",
+    "retention": "1 months",
+    "format": "<level>{level: <8}</level> <green>{time:YYYY-MM-DD HH:mm:ss.SSS}</green> request id: {extra[request_id]} - <cyan>{name}</cyan>:<cyan>{function}</cyan> - <level>{message}</level>"
+  }
+}
+```
+
+## Use
+Let's suppose the image is named **io** and we want to expose port 5000 in order to work. We can also image that we want to remove the container when it finishes, we can do this way:
```
Comment (flopezag) [inline docker/Readme.md:31]: Let's suppose the image is named **io** and we want to expose port 5000 in order to work. We can also imagine that we want to remove the container when it finishes, we can do it in the following way:

File: docker/Readme.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+```
+docker run --rm -p 5000:5000 --name io  iotagent-turtle
+```
+
+However, we might have Orion-ld somewhere else. So if we cant to connect our context broker somewhere, we can use this command:
```
Comment (flopezag) [inline docker/Readme.md:36]: However, we might have Orion-ld somewhere else. Therefore, if we can connect our context broker somewhere, we can use this command:

File: docker/Readme.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+```
+docker run --rm -p 5000:5000 --name io --add-host=orion-ld:192.168.1.206  iotagent-turtle
+```
+
+### Overriding config.json
+We could create our own config.json and override the one in the docker by default
```
Comment (flopezag) [inline docker/Readme.md:42]: We could create our own config.json and override the default configuration file in the docker:

File: docker/Readme.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+```
+docker run --rm -p 5000:5000 --name io -v our-local-config.json:/opt/IoTAgent-turtle/common/config.json iotagent-turtle
+```
+
+### As docker file
+
+We can consider writing a docker-compose.yaml file as the following one to start everything (orion-ld, the name of our orion server is named according to our config.json file:
```
Comment (flopezag) [inline docker/Readme.md:49]: We can consider writing a docker-compose.yaml file as the following one to start everything (orion-ld, the name of our orion server is named according to our config.json file, see "broker" key):

File: ngsild/ngsild_connector.py [hunk 9]
```diff
@@ -1,28 +1,8 @@
-#!/usr/bin/env python
```
Comment (flopezag) [inline ngsild/ngsild_connector.py:1]: why did you delete the header?

File: ngsild/ngsild_connector.py [hunk 10]
```diff
@@ -39,11 +19,29 @@ def get_url(self):
...
         d = d if type(d) is list else [d]
 
         for elem in d:
-            rc, r = c.send_data(json.dumps(elem))
+            try:
+                rc, r = self.send_data(json.dumps(elem))
+                return_info.append({"id": elem['id'],
+                                   "status_code": rc,
+                                   "reason": r})
+            except TypeError as e:
```
Comment (flopezag) [inline ngsild/ngsild_connector.py:32]: I suggest raising the exception due to in the server:parser we are controlling the different exceptions that we have in the server (lines 158-180).

I suggest creating a new 400 - 499 status code. I suggest using 422 Unprocessable Entity (status.HTTP_422_UNPROCESSABLE_ENTITY).

File: tests/common/test_commonclass.py [hunk 11]
```diff
@@ -55,3 +55,5 @@ def test_save(self):
             data = json.load(f)
         assert(data['id'] == urnid)
         assert(data['@context'] == context['@context'])
+
+    # TODO - Add tests with cclass.generate_id using update_id with a False value
```
Comment (flopezag) [inline tests/common/test_commonclass.py:59]: add new line

**Review by flopezag — CHANGES_REQUESTED** [review]

File: ngsild/ngsild_connector.py (hunk 9 above)
Comment (jicarretero) [reply ngsild/ngsild_connector.py:1]: I put it again.

File: docker/Readme.md (hunk 8 above)
Comment (jicarretero) [reply docker/Readme.md:49]: Changed the comment

File: docker/Readme.md (hunk 7 above)
Comment (jicarretero) [reply docker/Readme.md:42]: ok

File: docker/Readme.md (hunk 5 above)
Comment (jicarretero) [reply docker/Readme.md:31]: ok

File: docker/Readme.md (hunk 6 above)
Comment (jicarretero) [reply docker/Readme.md:36]: ok... changed in what I think is a better way.

File: docker/Readme.md (hunk 4 above)
Comment (jicarretero) [reply docker/Readme.md:12]: Ok

File: docker/Dockerfile (hunk 3 above)
Comment (jicarretero) [reply docker/Dockerfile:5]: Reset the labels in the final image for the docker

File: docker/Dockerfile (hunk 1 above)
Comment (jicarretero) [reply docker/Dockerfile:33]: Reset

File: docker/Dockerfile [hunk 12]
```diff
@@ -1,63 +1,44 @@
...
-ENV PATH="/usr/app/venv/bin:$PATH"
-CMD ["python", "agent.py", "server", "--host", "0.0.0.0", "--port", "5000"]
-HEALTHCHECK --interval=30s --timeout=30s --start-period=5s --retries=3 CMD curl -f https://localhost:${PORT}/version
+COPY config.json $INSTALLATION_PATH/common/config.json
+WORKDIR $INSTALLATION_PATH
 
 EXPOSE ${PORT}
+ENTRYPOINT /usr/local/bin/python agent.py server --host 0.0.0.0 --port ${PORT}
+
+expose $PORT
```
[the remark this answers is not in the document]
Comment (jicarretero) [reply docker/Dockerfile:44]: mistake

File: docker/Readme.md (hunk 2 above)
Comment (jicarretero) [reply docker/Readme.md:4]: Ok

File: docker/Dockerfile (hunk 1 above)
Comment (flopezag) [reply docker/Dockerfile:33]: Please check the comment because I have changed some of the values.

File: docker/Readme.md (hunk 2 above)
Comment (flopezag) [reply docker/Readme.md]: please check it again...

File: docker/Readme.md (hunk 8 above)
Comment (flopezag) [reply docker/Readme.md]: check "(see "broker" key)", You have two open parentheses but only one closing.

File: ngsild/ngsild_connector.py (hunk 9 above)
Comment (flopezag) [reply ngsild/ngsild_connector.py:1]: I do not see it

File: docker/Dockerfile (hunk 1 above)
Comment (jicarretero) [reply docker/Dockerfile:33]: OK

File: docker/Readme.md (hunk 2 above)
Comment (jicarretero) [reply docker/Readme.md]: Copy-pasted your "text"

File: docker/Readme.md (hunk 8 above)
Comment (jicarretero) [reply docker/Readme.md]: Ok

File: docker/Readme.md (hunk 6 above)
Comment (flopezag) [reply docker/Readme.md]: LGTM

File: ngsild/ngsild_connector.py (hunk 10 above)
Comment (flopezag) [reply ngsild/ngsild_connector.py:55]: LGTM

**Comment by flopezag** [comment]
LGTM
